File: flame/src/main/java/org/movieos/flame/utilities/DiscoveryService.java
package org.movieos.flame.utilities;

import android.content.Context;
import android.net.wifi.WifiManager;
import android.os.AsyncTask;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.text.format.Formatter;

import org.greenrobot.eventbus.EventBus;
import org.movieos.flame.models.FlameHost;
import timber.log.Timber;

import java.io.IOException;
import java.net.InetAddress;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.jmdns.JmDNS;
import javax.jmdns.ServiceEvent;
import javax.jmdns.ServiceInfo;
import javax.jmdns.ServiceListener;
import javax.jmdns.ServiceTypeListener;

public class DiscoveryService implements ServiceTypeListener, ServiceListener {

    private static final String FLAME_SERVICE_NAME = "_flame._tcp.local.";
    private static final int FLAME_SERVICE_PORT = 11812;

    private HashMap<String, ServiceEvent> mServices;

    private Context mContext;
    private JmDNS mJmDNSService;
    private WifiManager.MulticastLock mWifiLock;

    public DiscoveryService(Context context) {
        super();
        mContext = context.getApplicationContext();

        mServices = new HashMap<>();

        final WifiManager wifi = (android.net.wifi.WifiManager)mContext.getApplicationContext().getSystemService(android.content.Context.WIFI_SERVICE);
        mWifiLock = wifi.createMulticastLock("Flame");
        mWifiLock.setReferenceCounted(true);
        mWifiLock.acquire();

        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... params) {
                try {
                    String ip = Formatter.formatIpAddress(wifi.getConnectionInfo().getIpAddress());
                    InetAddress bindingAddress = InetAddress.getByName(ip);
                    mJmDNSService = JmDNS.create(bindingAddress, "Android");
                    mJmDNSService.addServiceTypeListener(DiscoveryService.this);

                    // broadcast that we are running flame
                    ServiceInfo serviceInfo = ServiceInfo.create(FLAME_SERVICE_NAME, "Flame", FLAME_SERVICE_PORT, mJmDNSService.getName());
                    mJmDNSService.registerService(serviceInfo);

                    Timber.v("Discovery started");
                } catch (IOException e) {
                    Timber.e(e, "Failed discovery");
                }
                return null;
            }
        }.execute();

        EventBus.getDefault().post(new HostsChangedNotification());
    }

    public void stop() {
        Timber.v("onDestroy");
        if (mWifiLock != null) {
            mWifiLock.release();
        }
        if (mJmDNSService != null) {
            mJmDNSService.unregisterAllServices();
            mJmDNSService.removeServiceTypeListener(this);
        }
    }

    /// data accessors

    @NonNull
    public List<FlameHost> getHosts() {
        return getHostsMatchingKey(null);
    }

    @NonNull
    public List<FlameHost> getHostsMatchingKey(String filter) {
        Map<String, List<ServiceEvent>> grouped = new HashMap<>();
        for (ServiceEvent service : mServices.values()) {
            String key = FlameHost.hostIdentifierForService(service);
            if (filter != null && !TextUtils.equals(filter, key)) {
                continue;
            }
            grouped.putIfAbsent(key, new ArrayList<>());
            grouped.get(key).add(service);
        }

        List<FlameHost> hosts = new ArrayList<>();
        for (final Map.Entry<String, List<ServiceEvent>> entry : grouped.entrySet()) {
            FlameHost host = new FlameHost(entry.getValue());
            hosts.add(host);
        }

        Collections.sort(hosts, (lhs, rhs) -> lhs.getTitle().compareToIgnoreCase(rhs.getTitle()));

        return hosts;
    }

    /// actions

    private String uniqueIdentifierForService(ServiceEvent service) {
        return String.format("%s %s", service.getName(), service.getType());
    }

    /// jmdns callbacks

    @Override
    public void serviceTypeAdded(ServiceEvent event) {
        //Timber.v("new type: " + event.getType());
        event.getDNS().addServiceListener(event.getType(), this);
    }

    @Override
    public void subTypeForServiceTypeAdded(ServiceEvent event) {
        //Timber.v("subTypeForServiceTypeAdded " + event);
    }

    @Override
    public void serviceAdded(ServiceEvent event) {
        //Timber.v("serviceAdded: " + event.getName() + " / " + event.getType());
        event.getDNS().requestServiceInfo(event.getType(), event.getName(), true);
    }

    @Override
    public void serviceRemoved(ServiceEvent event) {
        Timber.v("serviceRemoved: %s / %s", event.getName(), event.getType());
        final String key = uniqueIdentifierForService(event);
        if (mServices.containsKey(key)) {
            mServices.remove(key);
            EventBus.getDefault().post(new HostsChangedNotification());
        }
    }

    @Override
    public void serviceResolved(final ServiceEvent event) {
        Timber.v("serviceResolved: %s / %s", event.getName(), event.getType());
        mServices.put(uniqueIdentifierForService(event), event);
        EventBus.getDefault().post(new HostsChangedNotification());
    }

    public static class HostsChangedNotification {

    }

}

File: flame/src/main/java/org/movieos/flame/utilities/DiscoveryService.kt
package org.movieos.flame.utilities

import android.content.Context
import android.net.wifi.WifiManager
import android.os.AsyncTask
import android.text.TextUtils
import android.text.format.Formatter
import org.greenrobot.eventbus.EventBus
import org.movieos.flame.models.FlameHost
import timber.log.Timber
import java.io.IOException
import java.net.InetAddress
import java.util.*
import javax.jmdns.*

class DiscoveryService(tempContext: Context) : ServiceTypeListener, ServiceListener {

    private val services: HashMap<String, ServiceEvent> = HashMap()
    private val context: Context = tempContext.applicationContext
    private var jmDNSService: JmDNS? = null
    private val wifiLock: WifiManager.MulticastLock?

    init {
        val wifi = context.applicationContext.getSystemService(android.content.Context.WIFI_SERVICE) as android.net.wifi.WifiManager
        wifiLock = wifi.createMulticastLock("Flame")
        wifiLock.setReferenceCounted(true)
        wifiLock.acquire()

        object : AsyncTask<Void, Void, Void>() {
            override fun doInBackground(vararg params: Void): Void? {
                try {
                    val ip = Formatter.formatIpAddress(wifi.connectionInfo.ipAddress)
                    val bindingAddress = InetAddress.getByName(ip)
                    jmDNSService = JmDNS.create(bindingAddress, "Android")
                    jmDNSService!!.addServiceTypeListener(this@DiscoveryService)

                    // broadcast that we are running flame
                    val serviceInfo = ServiceInfo.create(FLAME_SERVICE_NAME, "Flame", FLAME_SERVICE_PORT, jmDNSService!!.name)
                    jmDNSService!!.registerService(serviceInfo)

                    Timber.v("Discovery started")
                } catch (e: IOException) {
                    Timber.e(e, "Failed discovery")
                }

                return null
            }
        }.execute()

        EventBus.getDefault().post(HostsChangedNotification())
    }

    fun stop() {
        Timber.v("onDestroy")
        wifiLock?.release()
        if (jmDNSService != null) {
            jmDNSService!!.unregisterAllServices()
            jmDNSService!!.removeServiceTypeListener(this)
        }
    }

    /// data accessors

    val hosts: List<FlameHost>
        get() = getHostsMatchingKey(null)

    fun getHostsMatchingKey(filter: String?): List<FlameHost> {
        val grouped = HashMap<String, MutableList<ServiceEvent>>()
        for (service in services.values) {
            val key = FlameHost.hostIdentifierForService(service)
            if (filter != null && !TextUtils.equals(filter, key)) {
                continue
            }
            if (grouped[key] == null) {
                grouped[key] = ArrayList()
            }
            grouped[key]?.add(service)
        }

        val hosts = ArrayList<FlameHost>()
        for ((_, value) in grouped) {
            val host = FlameHost(value)
            hosts.add(host)
        }

        Collections.sort(hosts) { lhs, rhs -> lhs.title.compareTo(rhs.title, ignoreCase = true) }

        return hosts
    }

    /// actions

    private fun uniqueIdentifierForService(service: ServiceEvent): String {
        return String.format("%s %s", service.name, service.type)
    }

    /// jmdns callbacks

    override fun serviceTypeAdded(event: ServiceEvent) {
        //Timber.v("new type: " + event.getType());
        event.dns.addServiceListener(event.type, this)
    }

    override fun subTypeForServiceTypeAdded(event: ServiceEvent) {
        //Timber.v("subTypeForServiceTypeAdded " + event);
    }

    override fun serviceAdded(event: ServiceEvent) {
        //Timber.v("serviceAdded: " + event.getName() + " / " + event.getType());
        event.dns.requestServiceInfo(event.type, event.name, true)
    }

    override fun serviceRemoved(event: ServiceEvent) {
        Timber.v("serviceRemoved: %s / %s", event.name, event.type)
        val key = uniqueIdentifierForService(event)
        if (services.containsKey(key)) {
            services.remove(key)
            EventBus.getDefault().post(HostsChangedNotification())
        }
    }

    override fun serviceResolved(event: ServiceEvent) {
        Timber.v("serviceResolved: ${event.name} / ${event.type}")
        services.put(uniqueIdentifierForService(event), event)
        EventBus.getDefault().post(HostsChangedNotification())
    }

    class HostsChangedNotification

    companion object {
        private val FLAME_SERVICE_NAME = "_flame._tcp.local."
        private val FLAME_SERVICE_PORT = 11812
    }

}
